### PR TITLE
Fix strong references to mutable objects in context.clone

### LINKED ISSUE #926: infer_stmts cannot infer multiple uses of the same AssignName
Given multiple assignments to the same target which both reference the same AssignName, infer_stmts fails for subsequent attempts after the first.

### Steps to reproduce

This appears to be a minimum working example, removing any part removes the effect:

```python
fails = astroid.extract_node("""
    pair = [1, 2]
    ex = pair[0]
    if 1 + 1 == 2:
        ex = pair[1]
    ex
""")
print(list(fails.infer()))
# [<Const.int l.2 at 0x...>, Uninferable]
```

For some context, I originally saw this with attributes on an imported module, i.e.

```python
import mod
ex = mod.One()
# later ... or in some branch
ex = mod.Two()
```

### Current behavior

See above.

### Expected behavior

Inlining the variable or switching to a different name works fine:

```python
works = astroid.extract_node("""
    # pair = [1, 2]
    ex = [1, 2][0]
    if 1 + 1 == 2:
        ex = [1, 2][1]
    ex
""")
print(list(works.infer()))
# [<Const.int l.3 at 0x...>, <Const.int l.5 at 0x...>]

works = astroid.extract_node("""
    first = [1, 2]
    second = [1, 2]
    ex = first[0]
    if 1 + 1 == 2:
        ex = second[1]
    ex
""")
print(list(works.infer()))
# [<Const.int l.2 at 0x...>, <Const.int l.3 at 0x...>]
```

I would expect that the first failing example would work similarly. This (only) worked
in astroid 2.5 and appears to have been "broken" by the revert of cc3bfc5 in 03d15b0 (astroid 2.5.1 and above).

### ``python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"`` output

```
$ python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"
2.5-dev
$ git rev-parse HEAD
03d15b0f32f7d7c9b2cb062b9321e531bd954344
```

### PR DESCRIPTION
**Pull request by nelfin**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

Since `InferenceContext.clone` passes `self.path` when
constructing a new instance, these two objects will hold references to
the same underlying set object. This caused modifications to
`context.path` further down in the stack to be reflected in "earlier"
objects and this would incorrectly detect a loop if the same node was
used twice in the inference of a some node (instead of only if an
inference result was used in inferring itself).

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|   | :sparkles: New feature |
|   | :hammer: Refactoring  |
|   | :scroll: Docs |

## Related Issue

Closes #926

### CODE REVIEW

File: tests/unittest_inference.py [hunk 1]
```diff
@@ -1703,7 +1703,8 @@ def __init__(self):
         """
         ast = extract_node(code, __name__)
         expr = ast.func.expr
-        self.assertIs(next(expr.infer()), util.Uninferable)
+        with pytest.raises(exceptions.InferenceError):
+            next(expr.infer())
```
Comment (nelfin) [inline tests/unittest_inference.py:1707]: This test has changed behaviour. I assume that it's okay for it to raise `InferenceError` instead of returning `Uninferable` as long as it still doesn't leak `StopIteration`

**Comment by nelfin** [comment]
Updated and rebased to appease code formatting overlords

File: tests/unittest_regrtest.py [hunk 2]
```diff
@@ -100,7 +100,7 @@ def test_numpy_crash(self):
         astroid = builder.string_build(data, __name__, __file__)
         callfunc = astroid.body[1].value.func
         inferred = callfunc.inferred()
-        self.assertEqual(len(inferred), 2)
+        self.assertEqual(len(inferred), 1)
```
[the remark this answers is not in the document]
Comment (hippo91) [reply tests/unittest_regrtest.py:103]: @nelfin. You are right. Can you please modify the `unittest_brain_numpy_core_umath.py` module, line 224 with:
```
                   self.assertTrue(
                    len(inferred_values) == 1,
                    msg="Too much inferred values ({}) for {:s}".format(
                        inferred_values, func_
                    ),
                )
```

**Review by hippo91 — APPROVED** [review]
@nelfin it is a great improvment. In fact you succeeded where i failed with #884. Congrats!
I just left a bunch of comments.
I also tried the tests of `pylint` with your astroid branch and one test is failing. It deals with logging module and it is possible that the warnings that are now emitted are correct. Could you please have a look to it? @Pierre-Sassoulas can you also have a look? I saw you just reintroduced the concerned lines, so you will be much pertinent than me.

File: tests/unittest_inference.py (hunk 1 above)
Comment (hippo91) [reply tests/unittest_inference.py:1707]: @cdce8p what is your opinion?

File: tests/unittest_inference.py [hunk 3]
```diff
@@ -5962,5 +5962,29 @@ def test_infer_list_of_uninferables_does_not_crash():
...
+def test_issue926_infer_stmts_referencing_same_name_is_not_uninferable():
+    code = """
+    pair = [1, 2]
+    ex = pair[0]
+    if 1 + 1 == 2:
+        ex = pair[1]
+    ex
+    """
+    node = extract_node(code)
+    inferred = list(node.infer())
```
[the remark this answers is not in the document]
Comment (nelfin) [reply tests/unittest_inference.py:5975]: Sorry, it was a habit I picked up. I'd originally written these tests with class instances but found the subscript as a simpler example. I had a note to myself that `list(node.infer())` was not equivalent to `node.inferred()` in all cases and to prefer the former (specifically, if `node` is an `Instance` then `node.inferred()` is `[<ClassDef ...>]` and `list(node.infer())` is `[<Instance ...>]` due to how `Proxy` forwards the call). I wanted the list of an instance and I couldn't figure out at the time if it was intended behaviour so I just made myself a note and moved on.

File: tests/unittest_inference.py [hunk 4]
```diff
@@ -5962,5 +5962,29 @@ def test_infer_list_of_uninferables_does_not_crash():
...
+    assert util.Uninferable not in inferred
```
[the remark this answers is not in the document]
Comment (nelfin) [reply tests/unittest_inference.py:5976]: Good point. Done.

File: tests/unittest_regrtest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nelfin) [reply tests/unittest_regrtest.py:103]: Done

File: tests/unittest_inference.py [hunk 5]
```diff
@@ -5962,5 +5962,29 @@ def test_infer_list_of_uninferables_does_not_crash():
     assert not inferred.elts
 
 
+def test_issue926_infer_stmts_referencing_same_name_is_not_uninferable():
```
[the remark this answers is not in the document]
Comment (nelfin) [reply tests/unittest_inference.py:5966]: Done

**Comment by nelfin** [comment]
> I also tried the tests of pylint with your astroid branch and one test is failing. It deals with logging module and it is possible that the warnings that are now emitted are correct. Could you please have a look to it?

I think the new messages are okay? There wasn't any warning previously raised for the following:
```python
class Logger(renamed_logging.Logger):
    pass

custom_logger = Logger('three')
custom_logger.info('testing {0}'.format('info'))  # here
custom_logger.info('testing %s' % 'info')   # and here
```
but those newly raised warning match the behaviour you would expect if `Logger` was being correctly inferred now. I'll update that test and submit a PR to pylint

File: tests/unittest_inference.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hippo91) [reply tests/unittest_inference.py:5976]: OK no pb!

File: tests/unittest_regrtest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hippo91) [reply tests/unittest_regrtest.py:103]: Thanks a lot!

**Comment by hippo91** [comment]
@nelfin i will wait for @Pierre-Sassoulas remarks concerning the pylint test before merging this. Thanks again.

**Comment by nelfin** [comment]
> @nelfin i will wait for @Pierre-Sassoulas remarks concerning the pylint test before merging this.

No problem.

> Thanks again.

Thank *you*!

File: tests/unittest_inference.py (hunk 1 above)
Comment (cdce8p) [reply tests/unittest_inference.py:1707]: I'm not too familiar with that part of the code base. If the pylint tests continue to pass, I would guess it's fine.

Comment (Pierre-Sassoulas) [reply tests/unittest_inference.py:1707]: Potentially this could cause new crash if an inference error is raised in calling code ? I don't think pylint tests cover all possible edge case.

**Comment by cdce8p** [comment]
@hippo91 I just looked at https://github.com/PyCQA/pylint/pull/4325 and also ran the other functional tests. The error with the logging tests is indeed a false-negative currently. Since all seems to work as expected then, I would say this is good to go.

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

**Review by cdce8p — CHANGES_REQUESTED** [review]
@Pierre-Sassoulas mentioned [here](https://github.com/PyCQA/astroid/pull/927#discussion_r610819685) that the change from `Uninferable` to `StopIteration` might have unintended consequences. After I saw it, I ran the changes against one of my repos and indeed saw multiple errors. Namely: `not-callable`, `unsupported-membership-test`, and `invalid-sequence-index`. I haven't investigated what code exactly is causing these, but it's safe to assume this shouldn't be merged just yet.

The Github Actions run: https://github.com/cdce8p/ha-core/runs/2308362444?check_suite_focus=true#step:7:58

**Comment by cdce8p** [comment]
I had some time to look at two of them and reduced it down to code snippets.

**unsupported-membership-test**
```py
# pylint: disable=missing-docstring,too-few-public-methods,invalid-name
from enum import Enum

class MyEnum(Enum):
    CONST1 = "const"

def name_in_enum(name):
    # false-positive: unsupported-membership-test
    if name in MyEnum.__members__:
        return
```

**not-callable**
```bash
pip install aiohttp python-telnet-vlc
```

```py
# pylint: disable=missing-docstring,unused-import,protected-access
from aiohttp import web
from python_telnet_vlc import VLCTelnet

def func():
    app = web.Application()
    app._router.freeze = lambda: None

class VlcDevice:
    def __init__(self):
        self._vlc = VLCTelnet("VLC-TELNET", "pwd", "0000")

    def media_previous_track(self):
        # False-positive: not-callable
        self._vlc.prev()

    def media_next_track_abc(self):
        # False-positive: not-callable
        self._vlc.next()
```

**Comment by nelfin** [comment]
> I had some time to look at two of them and reduced it down to code snippets. Maybe that helps.

Thanks for that, it helps a lot.

I spent some time trying to replicate these messages from your bulid log and debugging the causes. I picked up these errors from the log in the link:

```
Error: homeassistant/components/vlc_telnet/media_player.py:289:8: E1102: self._vlc.prev is not callable (not-callable)
Error: homeassistant/components/vlc_telnet/media_player.py:293:8: E1102: self._vlc.next is not callable (not-callable)
Error: homeassistant/components/mpd/media_player.py:393:14: E1102: self._client.next is not callable (not-callable)
Error: homeassistant/components/ezviz/binary_sensor.py:26:23: E1135: Value 'BinarySensorType.__members__' doesn't support membership test (unsupported-membership-test)
Error: homeassistant/components/ezviz/sensor.py:26:23: E1135: Value 'SensorType.__members__' doesn't support membership test (unsupported-membership-test)
Error: homeassistant/components/xmpp/notify.py:193:20: E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```

The `not-callable` messages stem from the fact that the previous behaviour was to infer `[<Const.None>, <BoundMethod.add>]` but after the application of this change it only infers `[Const.None]`. I haven't yet determined the cause and hopefully a smaller snippet will help.

The `Enum.__members__` seems to be a false-positive. I'll see why, given that it should probably be inferred as `['']` via `brain_namedtuple_enum`. I reduced the example to an almost identical snippet to yours.

The `invalid-sequence-index` from `xmpp/notify.py:193` is interesting. It is newly inferred, the type of `message` was not inferred before. Inferring `message` gets us an instance of `xmpp.Message`, but inferring the `Subscript` gets us a list since `bases.Instance.getitem` returns the first inferred result from `infer_call_result` on `__getitem__`. In the general case, we should probably update `infer_getitem` to return all values so that `pylint/safe_infer` doesn't return (maybe extend `_check_invalid_sequence_index` to check that all inferred values are sequence types, like how `no-member` only emits if none of the inferred values have that attribute). In the specific case of `xmpp.Message/ElementBase`, there's potential for a brain to be added to return Uninferable because the behaviour of `__getitem__` can change dynamically at runtime due to loaded plugins. Here is a minimal example that matches the behaviour that I consider to be a false-positive:

```python
class DynamicGetitem:
    def __getitem__(self, key):
        if key == 'attributes':
            return []
        return {'world': 123}


ex = DynamicGetitem()
ex['hello']['world']
#  E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```

**Comment by nelfin** [comment]
> The `Enum.__members__` seems to be a false-positive. I'll see why, given that it should probably be inferred as `['']` via `brain_namedtuple_enum`. I reduced the example to an almost identical snippet to yours.

This wasn't quite correct. `members = ['']` is for enums built via `Enum.__call__`, e.g. `Foo = Enum('Foo', 'bar baz')`.

So it looks like this has uncovered an existing bug: before the patch the value of `SensorType.__members__` was `Uninferable` so the membership test check was skipped, after the patch `__members__` is inferred to a `Property` and not transformed to the expected value of that property. Here's an example repro that fails with pylint 2.7.4 on current astroid master f2b197a4f8af0ceeddf435747a5c937c8632872a:

```python
# enum_repro.py

class BaseMeta(type):
    def __new__(metacls, cls, bases, classdict):
        classdict['__members__'] = metacls.__members__
        return super().__new__(metacls, cls, bases, classdict)

    @property
    def __members__(cls):
        return {'hello', 'world'}


class Parent(metaclass=BaseMeta):
    pass


class Derived(Parent):
    pass


assert 'hello' in Parent.__members__
assert 'world' in Derived.__members__
```

And example output from astroid:

```python
# enum_repro.py

import astroid
mod = astroid.MANAGER.ast_from_file('enum_repro_mod.py', 'mod', source=True)
Parent, Derived = mod.body[1], mod.body[2]
print(list(Parent.igetattr('__members__')))
# [<Set.set l.10 at 0x...>]
print(list(Derived.igetattr('__members__')))
# [<Property.__members__ l.8 at 0x...>]
```

**Comment by nelfin** [comment]
@cdce8p: `aiohttp` and `VLCTelnet` turned out to be red herrings. This case fails on current stable versions:

```python
class Example:
    def prev(self):
        pass
    def next(self):
        pass
    def other(self):
        pass


ex = Example()
ex.other()  # no warning
ex.prev()   # no warning
ex.next()   # no warning

import typing

ex.other()  # no warning
ex.prev()   # false-positive: not-callable
ex.next()   # false-positive: not-callable
```

**Comment by nelfin** [comment]
I've bisected this down to 78d5537b6a40a5d4f5f80bad7ba567ff716d728a. Pylint 2.3.1 passes this case with 20a7ae5de32f716c8d7a4f5bbcd071bf353df4b2 and fails with 78d5537b6a40a5d4f5f80bad7ba567ff716d728a

**Review by cdce8p — APPROVED** [review]
Ok, let's merge it
